Non-deterministic error count in multi-threaded mode

## Behavior with `typescript@6.0`

Type checking passes.

## Behavior with `tsgo`

`tsgo` produces non-deterministic error count in multi-threaded mode — module resolution races cause phantom `TS2307` errors

## Description

Running `tsgo --stableTypeOrdering` on a large TypeScript monorepo (~500+ packages) produces a **variable number of type errors on every run** with no code changes. Running the same check with `--singleThreaded` produces a **stable, lower baseline**.

This strongly suggests a race condition in module resolution or shared type state between goroutines.

## Environment

```
@typescript/native-preview: 7.0.0-dev.20260423.1
Platform: darwin arm64 (12 logical cores)
Monorepo: ~500 packages, yarn workspaces, moduleResolution: "bundler"
strict: true, skipLibCheck: false
```

---

## Steps to reproduce

Run the following command **5 times in succession** with no code changes:

```bash
tsgo --stableTypeOrdering
```

Then run once with:

```bash
tsgo --stableTypeOrdering --singleThreaded
```

## Observed behaviour

| Run | `--stableTypeOrdering` (multi-threaded) | `--stableTypeOrdering --singleThreaded` |
|-----|-----------------------------------------|-----------------------------------------|
| 1   | 47 errors                               | 12 errors                               |
| 2   | 85 errors                               | 12 errors                               |
| 3   | 48 errors                               | 12 errors                               |
| 4   | 45 errors                               | 12 errors                               |
| 5   | 59 errors                               | 12 errors                               |


## Expected behaviour

Error count should be deterministic across runs for the same input.

---

## Phantom error classes

Two distinct categories of errors appear and disappear non-deterministically:

### 1. Module resolution failures for packages that exist (`TS2307`)

These are the clearest evidence of a race — these packages exist on disk and resolve correctly in single-threaded mode:

```
src/features/payments/hooks/index.spec.ts(7,36):
  error TS2307: Cannot find module '@my-workspace/payments-core/hooks/' or its corresponding type declarations.

src/features/forms/storage/src/hooks/useFormCards.ts(10,8):
  error TS2307: Cannot find module '@my-workspace/forms-config/' or its corresponding type declarations.

src/features/wallet/components/Card/Card.web.tsx(5,29):
  error TS2307: Cannot find module '@my-workspace/wallet-components/' or its corresponding type declarations.
```

These are valid workspace packages. They resolve correctly 100% of the time under `--singleThreaded` and under `tsc`. They fail resolution only intermittently in multi-threaded mode, seemingly depending on goroutine scheduling?

### 2. Cascading type errors from failed resolution

When a module fails to resolve, downstream files see `any`-typed symbols and generate secondary `TS2322`/`TS2345` errors. These inflate the total error count significantly beyond the raw `TS2307`s.

---

## Hypothesis

The shared module resolution cache (or path-mapping lookup) has unsynchronised concurrent reads/writes. When goroutine A resolves package X while goroutine B simultaneously begins resolving a file that imports X, goroutine B may read a partial or empty cache entry and conclude the module does not exist. The outcome varies with goroutine scheduling, explaining the run-to-run variance.

`--singleThreaded` eliminates the race, producing stable and correct results.

---

## Workaround

```bash
tsgo --stableTypeOrdering --singleThreaded
```

Stable and correct output, but forfeits the parallelism speedup on large repos.

---

## Additional notes

- `--stableTypeOrdering` alone does not help — it addresses output ordering of union members, not concurrency
- The phantom errors are not reproducible with `tsc` under any circumstances

*Drafted with LLM assistance. All data and reproduction steps are from real observed behaviour.*

--- a/internal/module/resolver.go
+++ b/internal/module/resolver.go
@@ -1045,7 +1045,15 @@ func (r *resolutionState) loadModuleFromImmediateNodeModulesDirectory(extensions
 }
 
 func (r *resolutionState) loadModuleFromSpecificNodeModulesDirectory(ext extensions, moduleName string, nodeModulesDirectory string) *resolved {
-	candidate := tspath.NormalizePath(tspath.CombinePaths(nodeModulesDirectory, moduleName))
+	// Strip any trailing directory separator so that imports like `pkg/` and `pkg`
+	// produce identical `candidate` and `packageDirectory` strings. Otherwise the
+	// `package.json` info cache (which is keyed by normalized path but stores the
+	// caller's `PackageDirectory` verbatim) can hand back, under concurrent
+	// inserts, an entry whose `PackageDirectory` doesn't match `candidate`,
+	// causing `loadNodeModuleFromDirectoryWorker`'s `ComparePaths(candidate, ...)`
+	// check to fail and skip loading the package's `main`/`types` entry.
+	// https://github.com/microsoft/typescript-go/issues/3526
+	candidate := tspath.RemoveTrailingDirectorySeparator(tspath.NormalizePath(tspath.CombinePaths(nodeModulesDirectory, moduleName)))
 	packageName, rest := ParsePackageName(moduleName)
 	packageDirectory := tspath.CombinePaths(nodeModulesDirectory, packageName)
 	if packageName == "" {

--- a/internal/module/resolver_test.go
+++ b/internal/module/resolver_test.go
@@ -1,0 +1,153 @@
+package module_test
+
+import (
+	"strings"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/microsoft/typescript-go/internal/core"
+	"github.com/microsoft/typescript-go/internal/module"
+	"github.com/microsoft/typescript-go/internal/vfs"
+	"github.com/microsoft/typescript-go/internal/vfs/vfstest"
+)
+
+type resolutionHostStub struct {
+	fs  vfs.FS
+	cwd string
+}
+
+func (h *resolutionHostStub) FS() vfs.FS                  { return h.fs }
+func (h *resolutionHostStub) GetCurrentDirectory() string { return h.cwd }
+
+// Regression test for https://github.com/microsoft/typescript-go/issues/3526.
+//
+// Resolving a node_modules import with a trailing slash (e.g. `pkg/`) must
+// produce the same result as without one.
+func TestResolveModuleNameTrailingSlash(t *testing.T) {
+	t.Parallel()
+
+	fs := vfstest.FromMap(map[string]string{
+		"/repo/node_modules/pkg/package.json": `{"name":"pkg","main":"main.js","types":"main.d.ts"}`,
+		"/repo/node_modules/pkg/main.d.ts":    "export const x: number;",
+		"/repo/node_modules/pkg/main.js":      "exports.x = 1;",
+		"/repo/src/file.ts":                   "",
+	}, true)
+	host := &resolutionHostStub{fs: fs, cwd: "/repo"}
+	opts := &core.CompilerOptions{
+		ModuleResolution: core.ModuleResolutionKindBundler,
+		Module:           core.ModuleKindESNext,
+		Target:           core.ScriptTargetESNext,
+	}
+	resolver := module.NewResolver(host, opts, "", "")
+
+	for _, name := range []string{"pkg", "pkg/"} {
+		r, _ := resolver.ResolveModuleName(name, "/repo/src/file.ts", core.ModuleKindESNext, nil)
+		if !r.IsResolved() {
+			t.Errorf("%q failed to resolve", name)
+		}
+	}
+}
+
+// blockingFS wraps a vfs.FS and forces FileExists calls for `targetPath` to
+// block on `gate` until released. It also counts how many goroutines are
+// waiting at the gate. This is used to deterministically reproduce the
+// `package.json` info-cache insert race described in
+// https://github.com/microsoft/typescript-go/issues/3526.
+type blockingFS struct {
+	vfs.FS
+	targetPath string
+	gate       chan struct{}
+	waiting    atomic.Int32
+}
+
+func (f *blockingFS) FileExists(path string) bool {
+	if path == f.targetPath {
+		f.waiting.Add(1)
+		<-f.gate
+	}
+	return f.FS.FileExists(path)
+}
+
+// Regression test for https://github.com/microsoft/typescript-go/issues/3526.
+//
+// Two goroutines resolve the same package via specifiers that differ only by
+// a trailing slash (`pkg` and `pkg/`). A blocking FS holds both at the
+// `FileExists` check for `package.json` — *after* each has confirmed a
+// `package.json` info-cache miss but *before* either has called `Set`. When
+// released, both proceed to `LoadOrStore` and one of them loses. Without the
+// fix, the loser receives the winner's `InfoCacheEntry` whose
+// `PackageDirectory` doesn't match its own `candidate` (because one spelling
+// has a trailing slash and the other doesn't), and
+// `loadNodeModuleFromDirectoryWorker`'s `ComparePaths` check skips loading
+// the package's `main`/`types`. With no `index.*` present, resolution falls
+// through to "unresolved" — the phantom TS2307 the issue describes. This
+// test deterministically fails when the fix is reverted.
+func TestResolveModuleNameTrailingSlashRace(t *testing.T) {
+	t.Parallel()
+
+	const pkgJSONPath = "/repo/node_modules/pkg/package.json"
+	files := map[string]string{
+		// `types` points at a file that is not discoverable through any
+		// fallback path: there is no `index.*` and no `main`. The only way
+		// to resolve `pkg` (or `pkg/`) is via the package.json `types` field
+		// inside `loadNodeModuleFromDirectoryWorker`, which is exactly the
+		// step that the bug skips when `candidate` and
+		// `packageInfo.PackageDirectory` mismatch.
+		pkgJSONPath: `{"name":"pkg","types":"./typings/index.d.ts"}`,
+		"/repo/node_modules/pkg/typings/index.d.ts": "export const x: number;",
+		// Distinct containing files so each `ResolveModuleName` call has a
+		// unique module-resolution-cache key.
+		"/repo/src/a/file.ts": "",
+		"/repo/src/b/file.ts": "",
+	}
+	fs := &blockingFS{
+		FS:         vfstest.FromMap(files, true),
+		targetPath: pkgJSONPath,
+		gate:       make(chan struct{}),
+	}
+	host := &resolutionHostStub{fs: fs, cwd: "/repo"}
+	opts := &core.CompilerOptions{
+		ModuleResolution: core.ModuleResolutionKindBundler,
+		Module:           core.ModuleKindESNext,
+		Target:           core.ScriptTargetESNext,
+	}
+	resolver := module.NewResolver(host, opts, "", "")
+
+	type result struct {
+		name     string
+		resolved bool
+	}
+	results := make(chan result, 2)
+	var wg sync.WaitGroup
+	for _, name := range []string{"pkg", "pkg/"} {
+		containingFile := "/repo/src/a/file.ts"
+		if strings.HasSuffix(name, "/") {
+			containingFile = "/repo/src/b/file.ts"
+		}
+		wg.Go(func() {
+			r, _ := resolver.ResolveModuleName(name, containingFile, core.ModuleKindESNext, nil)
+			results <- result{name, r.IsResolved()}
+		})
+	}
+
+	// Wait for both goroutines to reach the FileExists gate, guaranteeing
+	// both have observed a package.json info-cache miss.
+	deadline := time.Now().Add(5 * time.Second)
+	for fs.waiting.Load() < 2 {
+		if time.Now().After(deadline) {
+			t.Fatalf("timed out waiting for both goroutines to reach FileExists gate; got %d", fs.waiting.Load())
+		}
+		time.Sleep(time.Millisecond)
+	}
+	close(fs.gate)
+
+	wg.Wait()
+	close(results)
+	for r := range results {
+		if !r.resolved {
+			t.Errorf("%q failed to resolve", r.name)
+		}
+	}
+}

